Serverless APM Documentation Update (#9384)
DataDog/documentation: content/en/serverless/distributed_tracing/_index.md

title: Distributed Tracing
kind: documentation
further_reading:
- link: "/tracing/"
  tag: "Documentation"
  text: "Explore Datadog APM"
- link: "/tracing/trace_search_and_analytics/#live-search-for-15-minutes"
  tag: "Documentation"
  text: "Live Search"
---

{{< img src="tracing/serverless_functions/ServerlessDistributedTrace.png" alt="Trace Serverless Functions"  style="width:100%;">}}

By connecting your serverless traces to metrics, Datadog provides a context-rich picture of your application’s performance, allowing you to better troubleshoot performance issues given the distributed nature of serverless applications. 

> The Datadog Python, Node.js, Ruby, Go and Java tracing libraries support distributed tracing for AWS Lambda. The easiest way to add tracing to your application is with the [Datadog Lambda Library][2], which includes the Datadog tracing library as a dependency.

Author: I'm going to add a note here for customers who are looking to get started quickly to go straight to the installation steps

## Choose your Tracing Library

{{< img src="integrations/amazon_lambda/lambda_tracing.png" alt="architecture diagram for tracing AWS Lambda with Datadog" >}}

To start using Datadog APM with your serverless application, you need to first choose between installing Datadog's tracing libraries (`dd-trace`) or AWS X-Ray tracing libraries based on your Lambda runtime and individual requirements. To see all of your traces in Datadog in real-time in the Live Search view, you need to use Datadog's tracing libraries.

| [Datadog APM with dd-trace][1]          | [Datadog APM with AWS X-Ray][2]           |
|---------------------------------|-------------------------------------------------------------------------|
| Uses Datadog APM's integration libraries for end-to-end tracing.  | Pulls traces from AWS X-Ray. |

> | Visualize your traces in Datadog in real-time. | Trace data available in Datadog after 2-5 minutes. |

Reviewer: Instead of mentioning `2-5` minutes, I would say a couple minutes -- it could be slower for customers with lots of traces to pull.

> | Tail-based sampling and fully customizable tag-based retention filters. | Sampling can be modified from the AWS X-Ray console. |

Reviewer: X-Ray sampling rules don't work for Lambda unless it has been fixed -- it's always sampled. https://docs.aws.amazon.com/lambda/latest/dg/services-xray.html

> | Tail-based sampling and fully customizable tag-based retention filters. | Sampling can be modified from the AWS X-Ray console. |

Reviewer: `Tail-based sampling and fully customizable tag-based retention filters.` => `Fully customizable tag-based retention rules.`? I think sampling and retention mean the same thing.

Author: They are slightly different to my understanding:

Sampling is what picks which traces are most important to index (tail-based sampling is desirable because it means that traces with errors/high latency, etc.) will always be indexed.

Retention filters is how long your traces are indexed for.

#### Go

The Datadog Lambda Library and tracing libraries for Go support:
- Manual correlation of Lambda logs and traces with trace ID and tag injection.
- Tracing HTTP requests invoking downstream Lambda functions or containers.
- Tracing dozens of additional out-of-the-box [Go][14] libraries.

You can trace your serverless functions in Datadog with [Datadog's tracing libraries][8] or by [installing our AWS X-Ray integration][12]. If you are using [Datadog's tracing libraries][8], and need to connect multiple Lambda function traces in event-driven architectures, we recommend you augment your traces by configuring _both_ AWS X-Ray and Datadog APM tracing libraries as described [here][10].

*Looking to trace through serverless resources not listed above? Open a feature request [here][9].*

#### Java

The Datadog Lambda Library for Java supports manual correlation of Lambda logs and AWS X-Ray traces with trace ID and tag injection. Datadog's tracing library for Java is currently in [beta][15]. We recommend you configure tracing for your Java serverless applications by [installing our AWS X-Ray integration][12].

> *Have feedback on the Datadog's tracing libraries for Java Lambda functions? Make sure to check out discussions going on in the <a href="https://datadoghq.slack.com/archives/CFDPB83M4">#serverless</a> channel in the <a href="https://chat.datadoghq.com/">Datadog Slack community</a>.*

Reviewer: Not sure if it's a big deal, but why not use the markdown link syntax instead of the html?

<div class="alert alert-info"> New to serverless monitoring? Follow the installation steps <a href="/serverless/installation/">here</a> to get started.</div>

To start using Datadog APM with your serverless application, you can choose between installing Datadog's tracing libraries (`dd-trace`) or AWS X-Ray tracing libraries based on your Lambda runtime and individual requirements. To see all of your traces in Datadog in real-time in the Live Search view, you need to use Datadog's tracing libraries.

| [Datadog APM with dd-trace][1]          | [Datadog APM with AWS X-Ray][2]           |
|---------------------------------|-------------------------------------------------------------------------|
| Uses Datadog APM's integration libraries for end-to-end tracing.  | Pulls traces from AWS X-Ray. |
| Visualize your traces in Datadog in real-time. | Trace data available in Datadog after a few minutes. |
| Tail-based sampling and fully customizable tag-based retention filters. | Sampling rate cannot be configured. |
| Support for Python, Node.js, Ruby, Go, Java. |  Support for all Lambda runtimes. |

> ### Runtime Recommendations

Author: ```suggestion
### Runtime recommendations
```

{{< partial name="serverless/serverless-apm-recommendations.html" >}}

#### Python and Node.js

The Datadog Lambda Library and tracing libraries for Python and Node.js support:
- Automatic correlation of Lambda logs and traces with trace ID and tag injection.
- Installation without any code changes using Serverless Framework, AWS SAM and AWS CDK integrations.
- Tracing HTTP requests invoking downstream Lambda functions or containers.
- Tracing synchronous Lambda invokes with the AWS SDK (coming soon for Node.js).
- Tracing asynchronous Lambda invocations through AWS SQS (coming soon for Node.js).
- Tracing dozens of additional out-of-the-box [Python][7] and [Node.js][11] libraries.

> For Python and Node.js serverless applications, we recommend you [install Datadog's tracing libraries][8]. If your application requires AWS X-Ray active tracing in AWS managed services such as API Gateway or Step Functions, we recommend you augment AWS X-Ray traces with Datadog APM by configuring _both_ AWS X-Ray and Datadog APM tracing libraries as described [here][10].

Reviewer 2: ```suggestion
For Python and Node.js serverless applications, Datadog recommends you [install Datadog's tracing libraries][8]. If your application requires AWS X-Ray active tracing in AWS managed services such as API Gateway or Step Functions, we recommend you augment AWS X-Ray traces with Datadog APM by configuring _both_ AWS X-Ray and Datadog APM tracing libraries as described [here][10].
```

- Tracing HTTP requests invoking downstream Lambda functions or containers.
- Tracing dozens of additional out-of-the-box [Go][14] libraries.

You can trace your serverless functions in Datadog with [Datadog's tracing libraries][8] or by [installing our AWS X-Ray integration][12]. If you are using [Datadog's tracing libraries][8], and need to connect multiple Lambda function traces in event-driven architectures, we recommend you augment your traces by configuring _both_ AWS X-Ray and Datadog APM tracing libraries as described [here][10].

*Looking to trace through serverless resources not listed above? Open a feature request [here][9].*

#### Java

The Datadog Lambda Library for Java supports manual correlation of Lambda logs and AWS X-Ray traces with trace ID and tag injection. Datadog's tracing library for Java is currently in [beta][15]. We recommend you configure tracing for your Java serverless applications by [installing our AWS X-Ray integration][12].

*Have feedback on the Datadog's tracing libraries for Java Lambda functions? Make sure to check out discussions going on in the [#serverless][18] channel in the [Datadog Slack community][19].*

#### .NET

> We recommend you configure tracing for your .NET AWS Lambda serverless applications by [installing our AWS X-Ray integration][12].

Reviewer 2: ```suggestion
Datadog recommends you configure tracing for your .NET AWS Lambda serverless applications by [installing our AWS X-Ray integration][12].
```

Learn more about tracing through .NET Azure serverless applications [here][16].

*Looking to trace through .NET AWS Lambda serverless applications with Datadog tracing libraries? Open a feature request [here][9].*

> ### Hybrid Environments

Reviewer 2: ```suggestion
### Hybrid environments
```
these should all be sentence case (anything that's not an h1)
